modules/lib/xml.nix

```diff
@@ -0,0 +1,36 @@
+{ libxslt, libxml2, runCommand }:
+
+rec {
+  genXMLFile = input:
+    runCommand "generated-xml" {
+      input = builtins.toXML input;
+
+      prefix = ''
+        <?xml version="1.0" encoding="utf-8"?>
+      '' + (if input ? doctype then
+        "<!DOCTYPE ${input.root.name} ${input.doctype}>"
+      else
+        "");
+
+      stylesheet = builtins.toFile "stylesheet.xsl" ''
+        <?xml version='1.0' encoding='UTF-8'?>
+        <xsl:stylesheet xmlns:xsl='http://www.w3.org/1999/XSL/Transform' version='1.0'>
+          <xsl:output method="xml" encoding="utf-8" omit-xml-declaration="yes" />
+          <xsl:template match='attrs[attr[@name="name"]]'>
+            <xsl:element name='{attr[@name="name"]/string/@value}'>
+              <xsl:for-each select='attr[@name="attrs"]/attrs/*[string]'>
+                <xsl:attribute name='{@name}'>
+                  <xsl:value-of select='string/@value' />
+                </xsl:attribute>
+              </xsl:for-each>
+              <xsl:apply-templates select='attr[@name="children"]/list/*' />
+              <xsl:value-of select='attr[@name="content"]/string/@value' />
+            </xsl:element>
+          </xsl:template>
+        </xsl:stylesheet>
+      '';
+    } ''
+      (echo "$prefix"; echo "$input" | ${libxslt}/bin/xsltproc $stylesheet -) | ${libxml2}/bin/xmllint --format - > $out
```

Comment:
Why do you need a subshell for this command?

Reply:
This pseudocode represents the idea here:

```python
processed = xsltproc(input, stylesheet)
prefixed = prefix + processed
formatted = xmllint.format(prefixed)
out.write(formatted)
```
There might be a simpler way to do this, it's been a few months since I've looked at this code.

Reply:
Ah okay, I missed the fact that the two `echo`s needed to be piped. A more explicit way to write this would be like the following:

```bash
       {echo "$prefix"; echo "$input";} | ${libxslt}/bin/xsltproc $stylesheet - | ${libxml2}/bin/xmllint --format - > $out 
```

As it is written, the `prefix` isn't getting piped into `xsltproc`, only in `xmllnt` (is it on purpose?).

Reply:
This was on purpose. I wasn't able to get `xsltproc` to properly emit a doctype. However, testing now, that works perfectly. I'm not sure why I was having trouble before.